improve documentation, add doctests, fix bugs found on the way

ericphanson opened the pull request:
I realized that after #8 some docstrings I didn't want to include (e.g. `_findmin`, which hopefully will be gone soon, ref #9) were there, and others didn't exist (e.g. `match` and `explain`). So here I've tried to fix that.

ericphanson committed:
fix up docs

ericphanson committed:
Move docstring to after `Corpus` defined

hannahilea commented on src/printing.jl:
````diff
@@ -125,6 +125,38 @@ function printcontext(io::IO, m::QueryMatch; context=40)
+
+```
+"""
+explain
````
This `explain` kinda comes out of nowhere, eh?

hannahilea commented on src/core.jl:
```diff
@@ -75,6 +75,23 @@ struct Query <: AbstractQuery
+Looks for a match for `query` in the text. Returns either `nothing`
+if no match is found, or a [`QueryMatch`](@ref) object.
+"""
+Base.match(query::AbstractQuery, text::Union{Document,Corpus})
```
Maybe I'm missing something obvious, but where's the function definition?

ericphanson replied:
Yeah, so in all these I'm on-purpose not attaching the docstring to the specific methods, because the existing methods are more narrow than where I want the docstring to apply. E.g. here I want the docstring to apply to any method of `Base.match` that fits that signature (instead of sticking it on just one of the `match` methods, or copying it by hand to all of them). For the explain one, I want it to apply to any method of `explain`. 

Why does it matter? if for example you do `? +` then you get all the docstrings for any method of `+`. But if you do `? +(1,1)`, you get just 1 docstring, the most specific one that applies to those arguments.

hannahilea replied:
😮  TIL!

hannahilea commented on src/core.jl:
```diff
@@ -75,6 +75,23 @@ struct Query <: AbstractQuery
+Looks for all matches for `query` in the text. Returns either a
+`Vector` `QueryMatch` objects corresponding to all of the matches found.
+"""
+match_all(query::AbstractQuery, text::Union{Document,Corpus})
```
ditto

ericphanson commented:
preview is at https://beacon-biosignals.github.io/KeywordSearch.jl/previews/PR11/, which you can get from clicking the "details" button in the "documenter/deploy" check

ericphanson committed:
fix bug: newlines need to be processed out for `word_boundary` to work

ericphanson committed:
add missing methods: `match_all` for NamedQueries

ericphanson committed:
Switch printing tests to be doctests, add doctests to tests

ericphanson committed:
use more doctests and improve docstrings

ericphanson committed:
organize API documentation

ericphanson committed:
Fix doctest

ericphanson committed:
add `fix_doctests.jl` script

ericphanson committed:
rm `using KeywordSearch` from doctests

ericphanson committed:
format

ericphanson committed:
fix backticks

ericphanson committed:
fix printing

ericphanson committed:
fix doctest -- very sensitive!

ericphanson committed:
last tiny tweaks

ericphanson commented:
Marking as draft since I want to do some more doc changes

ericphanson commented:
Ok, I made a bunch of changes, which I tried to group into clear commits. I found two bugs when improving the docs:
*  `fix_punct` wasn't removing `\n`'s, which meant the test in the docstring for `word_boundary` wasn't passing-- but since it was just a docstring instead of a doctest, I didn't know. I changed `fix_punct` to catch this, and made the test into a doctest
* `match_all` wasn't defined for `NamedQuery`s, so I added that and added tests

I decided to embrace doctests, in particular for testing printing, because it's easy to update them if the printing changes (see `docs/fix_doctests.jl`. I've added a bunch, and added them to `runtests.jl` so that they get tested as part of CI.

I think the codecov drop is spurious; the lines it says that are not covered are line 31-34 in printing.jl (https://app.codecov.io/gh/beacon-biosignals/KeywordSearch.jl/compare/11/changes#D1L31), but those are covered by the doctests for the show method for `Or` etc. I think maybe that `print_tree_rstrip` function is getting inlined and then codecov doesn't see coverage for it?

hannahilea commented on test/runtests.jl:
```diff
@@ -15,6 +15,12 @@ function with_replacements(f, replaces...)
+@testset "Doctests" begin
+    DocMeta.setdocmeta!(KeywordSearch, :DocTestSetup, :(using KeywordSearch);
+                        recursive=true)
+    Documenter.doctest(KeywordSearch)
```
Oh, cool.

hannahilea reviewed (approved):
I didn't know about `Documenter.doctest`! Niiice.

ericphanson commented:
I rebased after merging the others and fixed an issue with the printing. Not totally sure what happened, but I think doctests are very sensitive to whitespace, i.e. you need to start with the `julia>` line right after the ```````` ```jldoctest ```````` line.

I also learned from https://discourse.julialang.org/t/show-and-showcompact-on-custom-types/8493/4 that 3-argument show is preferred over checking `compact` as I was doing-- that's good to know, because `compact` wasn't triggering for arrays, only things like `Set`s An array of `Corpus`s will use the 2-argument show to print the entries, but a corpus alone will use the 3-argument one, so I use a 1-liner in the 2-argument version and add more details in the 3-argument one.
